### PR TITLE
panel/plugin: Fix settings storing/restoring

### LINKED ISSUE #1741: The problem with Reset button
This issue may not be limited to the panel but it can be seen in the config dialogs of some of its plugins:

The Rest button doesn't do a complete job everywhere. Sometimes, it doesn't change the states of widgets; sometimes it doesn't cause an on-the-fly savings of settings.

EDIT: The rest of this comment was a wrong description of the cause and created confusion. Sorry for that!

~In some places, when the user changes a setting, the new value is immediately saved. Apart from the problem that this can cause in the case of widgets like spinbox (the user turns the mouse wheel and the config file is overwritten constantly), after it happens, the reset button may lose its functionality because it reads the config file and shows the new values instead of the old ones~.

~In all config dialogs, the new values should be saved only when their OK or Apply buttons are pressed — or, at least, when their Close buttons are pressed, in case there's no OK/Apply button. Only in that way, the Reset button makes sense. "Reset" means "Restore Old Values".~

### PR DESCRIPTION
**Pull request by palinek**
- Add emitting settingsChanged() into PluginSettings::loadFromCache() to
  notify plugin/users, that settings were updated. Previously the signal
  was probably emitted by the underlying object detecting the settings
  file changed, but this is not the case any more (probably enhancement
  in the underlying Qt's QSettings optimization).

- Add PlutingSettings::storeToCache() for the plugin configure dialog to
  be able to store to cache on closing the dialog -> after re-showing or
  showing freshly created one, we expect to be able to reset to the
  values shown initialy in this dialog.

- Delete plugin's config dialog before deleting the plugin object
  itself, as the config dialog can rely/use other plugin objects and
  deleting/closing the dialog late after plugin object destroyed can
  cause problems (invalid memory access).

closes #1741

### CODE REVIEW

**Comment by tsujan** [comment]
This isn't needed. It complicates the code unnecessarily. Please read my replies at https://github.com/lxqt/lxqt-panel/pull/1744

File: panel/lxqtpanelpluginconfigdialog.cpp [hunk 1]
```diff
@@ -58,6 +58,15 @@ PluginSettings& LXQtPanelPluginConfigDialog::settings() const
...
 
+/************************************************
+
+ ************************************************/
+void LXQtPanelPluginConfigDialog::closeEvent(QCloseEvent *event)
+{
+    mSettings.storeToCache();
+    return QDialog::closeEvent(event);
+}
+
```
[the remark this answers is not in the document]
Comment (tsujan) [reply panel/lxqtpanelpluginconfigdialog.cpp:69]: Apart from not being needed, this is also at odd with the on-the-fly applying of settings change.

[the remark this answers is not in the document]
Comment (tsujan) [reply panel/lxqtpanelpluginconfigdialog.cpp:69]: @palinek, please forget about what I wrote at https://github.com/lxqt/lxqt-panel/issues/1741#issue-1167305638. As I said at https://github.com/lxqt/lxqt-panel/issues/1741#issuecomment-1065971820, I remembered everything later.

[the remark this answers is not in the document]
Comment (palinek) [reply panel/lxqtpanelpluginconfigdialog.cpp:69]: > Apart from not being needed, this is also at odd with the on-the-fly applying of settings change.

?
Aren't you misreading the code? This does:
- when the config dialog is closed, the current (and applied) settings are stored into "old values cache"
- when the config dialog is opened next time, the reset button will be (optionally) resetting to values "from previous close" (or from the plugin creation if no config showed yet)

[the remark this answers is not in the document]
Comment (tsujan) [reply panel/lxqtpanelpluginconfigdialog.cpp:69]: OK but it isn't needed. We don't need to be worried about the cache; it'll be correct if Reset does its job.

**Comment by palinek** [comment]
Updated... re-added emitting the `settingsChanged()` again... as this does the trick in case lxqt-panel is used with "default" configuration (no configuration file specified in command line options).

**Comment by palinek** [comment]
Heh.... after next testing -> the settings cache is messing up the config file. There is some deeper problem, here is the difference between pre-existing config file and after the reset button is pressed:
```diff
--- ../test_panel.conf  2022-03-14 13:32:47.044289979 +0100
+++ ../test1.conf       2022-03-14 13:32:10.396083054 +0100
@@ -4,8 +4,35 @@
 panels=panelTest
 
 [mainmenuTest]
+__userfile__=true
 alignment=Left
-ownIcon=false
+iconTheme=
+mainmenuTest\alignment=Left
+mainmenuTest\ownIcon=true
+mainmenuTest\showText=true
+mainmenuTest\type=mainmenu
+ownIcon=true
+panelTest\alignment=0
+panelTest\animation-duration=500
+panelTest\background-color=@Variant(\0\0\0\x43\0\xff\xff\0\0\0\0\0\0\0\0)
+panelTest\background-image=
+panelTest\desktop=0
+panelTest\font-color=@Variant(\0\0\0\x43\0\xff\xff\0\0\0\0\0\0\0\0)
+panelTest\hidable=false
+panelTest\hide-on-overlap=false
+panelTest\iconSize=32
+panelTest\lineCount=2
+panelTest\lockPanel=false
+panelTest\opacity=100
+panelTest\panelSize=80
+panelTest\plugins=mainmenuTest
+panelTest\position=Right
+panelTest\reserve-space=false
+panelTest\show-delay=500
+panelTest\visible-margin=false
+panelTest\width=80
+panelTest\width-percent=true
+panels=panelTest
 showText=true
 type=mainmenu
 
```

**Comment by tsujan** [comment]
> here is the difference between pre-existing config file and after the reset button is pressed

Yes, there is, without this patch too.

**Comment by tsujan** [comment]
The problem doesn't affect the functionality but makes a mess in `~/.config/lxqt/panel.conf`.

**Comment by tsujan** [comment]
It's in `PluginSettings::loadFromCache()`.

**Comment by tsujan** [comment]
The bug was in liblxqt. → https://github.com/lxqt/liblxqt/pull/306

**Comment by tsujan** [comment]
How? I can't reproduce the bug with https://github.com/lxqt/liblxqt/pull/306 — and without this patch.

**Comment by palinek** [comment]
Try to create `PluginSettings` object, call `PluginSettings::loadFromCache()`. This should make no difference in the underneath config file. But the `SettingsCache` is constructed wrongly in `PluginSettings` construction time (there are all the keys in from the config file stored into the cache), which will result in flushing all config keys under our plugin section ...

**Comment by tsujan** [comment]
> which will result in flushing all config keys under our plugin section

Yes, but

 1, It has no effect (after https://github.com/lxqt/liblxqt/pull/306) and
 2.  Why should loading from cache save settings? After all, it's just about loading. Also, in a future plugin, the coder may not want to save settings at this stage, for whatever reason, e.g., he may want/need to add Apply and OK buttons.

**Comment by tsujan** [comment]
If there's no OK/Apply button, yes, because the user knows settings should be applied on-the-fly. With OK and Apply buttons, we don't want to save anything before pressing them. Are we prohibiting the usage of OK/Apply buttons here? Sorry if I've misread the code — I'm very busy with other codes now — but I ask because I want to be sure.

**Comment by palinek** [comment]
> With OK and Apply buttons, we don't want to save anything before pressing them. Are we prohibiting the usage of OK/Apply buttons here? 

No we're not prohibiting anything, but encouraging the unified on-they-fly updating. If the plugin wants to force ok/apply/cancel it needs to implement it by itself.

**Comment by tsujan** [comment]
> Yes, because there is a sync() call luckily in between (as I said previously).

Sorry, however I look at it, I can't find a problem with `sync()`. Maybe I should come to this later...

**Comment by tsujan** [comment]
OK, coming back to this — free from other codes.

It doesn't prevent a "sleeping bug" but fixes an *existing bug*, which isn't related to what is fixed in https://github.com/lxqt/liblxqt/pull/306. Previously (a few hours ago), it also worked around the problem that's now fixed in https://github.com/lxqt/liblxqt/pull/306. That was the source of my confusion: the workaround isn't needed anymore but the fix is.

What does it fix? The settings cache might not be correct the second time a config dialog was opened. This patch guarantees a correct cache. This isn't a future bug but an existing one ;)

It addition, it discourages Apply/OK buttons (I don't like that but it's beside the point). Three changes in a single PR (actually 4)!

Feel free to merge it as it is. I'll need it to be merged for fixing https://github.com/lxqt/lxqt-panel/issues/1741 and changing its previous fixes according to it (perhaps, in a single PR).

**Comment by tsujan** [comment]
It's OK. I'll follow you in future ;) Who has time for several PRs?

**Review by tsujan — APPROVED** [review]

**Comment by tsujan** [comment]
> Should I wait for stefonarch to test this before merging?

I don't think he'll find a difference, because of the PRs that have been already merged (and I should change them after this is merged).

**Comment by palinek** [comment]
OK... merged. Thanks for reviewing.

**Comment by tsujan** [comment]
I'd seen the problem before but, because of it, I backed up my panel's conf file in its best state. That, in turn, made me forget about the problem (I restored the backup without taking a look at the config file).

**Comment by stefonarch** [comment]
Reconfigured all, now it makes sense looking at it. Excerpt  from the old file:
```
mount2\type=mount
mount2\volume\alignment=Right
mount2\volume\device=0
mount2\volume\type=volume
mount2\worldclock\alignment=Right
mount2\worldclock\autoRotate=true
mount2\worldclock\customFormat="dddd, d MMM |<b><big> H:mm</big> </b>"
mount2\worldclock\dateFormatType=custom
mount2\worldclock\dateLongNames=false
mount2\worldclock\datePadDay=false
mount2\worldclock\datePosition=below
mount2\worldclock\dateShowDoW=true
mount2\worldclock\dateShowYear=false
mount2\worldclock\defaultTimeZone=Europe/Rome
```
``

**Comment by tsujan** [comment]
Yes, that's prevented both in https://github.com/lxqt/liblxqt/pull/306 and here. The main usefulness of this PR was elsewhere.

**Comment by tsujan** [comment]
It's hard to describe but, in short, the PR guarantees remembering of the old settings the next time you open the config dialog (of Taskbar). The rest isn't so visible to the user.

**Pull request merged**